test/testsimplifytemplate.cpp

```diff
@@ -5368,6 +5369,73 @@ class TestSimplifyTemplate : public TestFixture {
         }
     }
 
+    void templateTypeDeduction5() {
+        {
+            const char code[] = "class Fred {\n"
+                                "public:\n"
+                                "    template <class T> Fred(T t) { }\n"
+                                "};\n"
+                                "Fred fred1 = Fred(0);\n"
+                                "Fred fred2 = Fred(0.0);\n"
+                                "Fred fred3 = Fred(\"zero\");\n"
+                                "Fred fred4 = Fred(false);";
+            const char exp[]  = "class Fred { "
+                                "public: "
+                                "Fred<int> ( int t ) ; "
+                                "Fred<double> ( double t ) ; "
+                                "Fred<constchar*> ( const char * t ) ; "
+                                "Fred<bool> ( bool t ) ; "
+                                "} ; "
+                                "Fred fred1 ; fred1 = Fred<int> ( 0 ) ; "
+                                "Fred fred2 ; fred2 = Fred<double> ( 0.0 ) ; "
+                                "Fred fred3 ; fred3 = Fred<constchar*> ( \"zero\" ) ; "
+                                "Fred fred4 ; fred4 = Fred<bool> ( false ) ; "
+                                "Fred :: Fred<int> ( int t ) { } "
+                                "Fred :: Fred<double> ( double t ) { } "
+                                "Fred :: Fred<constchar*> ( const char * t ) { } "
+                                "Fred :: Fred<bool> ( bool t ) { }";
+            ASSERT_EQUALS(exp, tok(code));
+        }
+        {
+            const char code[] = "namespace NS {\n"
+                                "class Fred {\n"
+                                "public:\n"
+                                "    template <class T> Fred(T t) { }\n"
+                                "};\n"
+                                "Fred fred1 = Fred(0);\n"
```

Comment:
What about passing a variable such as `int i;Fred fred1 = Fred(i);`?

Reply:
Type deduction currently only supports literals. See https://github.com/danmar/cppcheck/commit/c4b3d4cd772b3bdc9d4b40e66133000c7e5a60f7

This is just another baby step in a marathon.  I fixes only one of the many bugs needed to properly fix #9983.

Reply:
Make that #9985.

Reply:
The template simplifier runs very early in the tokenizer (before setVarid) so it's not easy to get the type of a variable.  That's why only literals are supported now.  It should be possible to get the type from Foo(new int).

Reply:
I see, it seems like we need to run the TemplateSimplifier before and after the SymbolDatabase. Run a cycle of TemplateSimplifier->SymbolDatabase->TemplateSimplifier->SymbolDatabase, but I dont know if thats possible without changes to TemplateSimplifier or SymbolDatabase.

Reply:
We could run setVarId before the TemplateSimplifier and again afterwards but I'm not sure if it's designed to run more than once.

Reply:
There is a self test error:
```
lib/valueptr.h:44:0: warning: Class 'ValuePtr' has a constructor with 1 argument that is not explicit. [noExplicitConstructor]
    template <class U>

```
but there is already a suppression for it in the code.
```
    template <class U>
    // cppcheck-suppress noExplicitConstructor
    ValuePtr(const U& value) : mPtr(cloner<U>::apply(&value)), mClone(&cloner<U>::apply)
    {}
```

Reply:
> We could run setVarId before the TemplateSimplifier and again afterwards but I'm not sure if it's designed to run more than once.

You can do it. But it will clear all varids and set all varids again. It is not optimal. But anyway it seems better that we are correct than fast.

Reply:
Varid wont be sufficient to get all type deduction as it would miss types for arbitrary expressions and library functions and types. It seems it would be better to use `ValueType` as this already computes the type information. I guess this would require us to remove the AST and SymbolDatabase and recreate it.

Reply:
I think the easiest thing to do is only run setVarId() the first time we encounter a template function which requires argument type deduction from a variable.  The template simplifier would then know when to clear everything. It may need to run setVarId() multiple times with nested templates. We need to fix what we have now before adding this new capability.

I get the explicit warning on on line 44 of valueptr.h with and without this patch so the suppression in the code doesn't seem to be working.

Reply:
> I think the easiest thing to do is only run setVarId() the first time we encounter a template function which requires argument type deduction from a variable. 

That would allow us to deduce types from non-auto variables, but how would we deduce the type for expressions(ie `f(a + 1)` or `f(g())`)? It seems like we need `ValueType` and the AST for that.

Reply:
I understand what you are saying but we can't do that with the current design and I'm not sure how we can incrementally get to what you want. We would need to rewrite the tokenizer, template simplifier and symbol database.

Reply:
I understand why we would need to make changes to the template simplifier. I dont understand why we would need to make changes to the SymbolDatabase. The SymbolDatabase seems to run fairly independent of the template simplifier.

Reply:
I tried using the astutils function getArguments within the template simplifier function getFunctionArguments and it didn't work.

This code:
```
int a = 1;
template <typename T> void f(T x, T y) { a = x + y; }
void test() { f(0, 0); }
```
produced this debug info:
```
function: "f":3:15 
0 astArgs: 
2 args: "0":3:17 "0":3:20 
function: "f":2:28 
0 astArgs: 
2 args: "T":2:30 "T":2:35 
```

Can this patch go in now as is?  This patch has nothing to do with template argument type deduction. It just fixes finding template constructors.

Reply:
Not sure how to interpret your debug output. The `getArguments` should work for the function call as far as I know. But not for the function declaration.

Reply:
>  The getArguments should work for the function call as far as I know.

`getArguments` needs the AST, which hasn't been created yet.

Reply:
Adding createAst give me the arguments.  I'm running it for each pass of the template simplifier because it's modifying the token list.  How do I ensure that the ast is in sync with the token list as I am editing it?

Reply:
> How do I ensure that the ast is in sync with the token list as I am editing it?

That is a good question. I guess either if you know what you are doing you could just tweak the astOperands and astParents a bit in the template simplifier.

But you could also clear the old AST for the statement and then run `createAstAtToken` somehow.

Reply:
The AST has problems with templates.  This patch should function identically because I'm just using getArguments() for function calls and I works but testrunner fails quickly in validateAst from infinite recursion.

```
TestAutoVariables::returnReference15
terminate called after throwing an instance of 'InternalError'
make: *** [Makefile:304: test] Aborted (core dumped)
```

```
diff --git a/lib/templatesimplifier.cpp b/lib/templatesimplifier.cpp
index 36514601b..6a18c21e6 100644
--- a/lib/templatesimplifier.cpp
+++ b/lib/templatesimplifier.cpp
@@ -24,6 +24,7 @@
 #include "token.h"
 #include "tokenize.h"
 #include "tokenlist.h"
+#include "astutils.h"
 
 #include <algorithm>
 #include <cassert>
@@ -771,7 +772,7 @@ void TemplateSimplifier::addInstantiation(Token *token, const std::string &scope
         mTemplateInstantiations.emplace_back(instantiation);
 }
 
-static void getFunctionArguments(const Token *nameToken, std::vector<const Token *> &args)
+static void getTemplateArguments(const Token *nameToken, std::vector<const Token *> &args)
 {
     const Token *argToken = nameToken->tokAt(2);
 
@@ -872,8 +873,7 @@ void TemplateSimplifier::getTemplateInstantiations()
 
             // look for function instantiation with type deduction
             if (tok->strAt(1) == "(") {
-                std::vector<const Token *> instantiationArgs;
-                getFunctionArguments(tok, instantiationArgs);
+                std::vector<const Token *> instantiationArgs = getArguments(tok);
 
                 std::string fullName;
                 if (!qualification.empty())
@@ -897,7 +897,7 @@ void TemplateSimplifier::getTemplateInstantiations()
                             continue;
 
                         std::vector<const Token *> declarationParams;
-                        getFunctionArguments(pos->second->nameToken(), declarationParams);
+                        getTemplateArguments(pos->second->nameToken(), declarationParams);
 
                         // function argument counts must match
                         if (instantiationArgs.empty() || instantiationArgs.size() != declarationParams.size())
@@ -3704,6 +3704,16 @@ void TemplateSimplifier::simplifyTemplates(
         // Locate user defined partial specializations.
         getPartialSpecializations();
 
+        if (passCount) {
+            for (Token *tok = mTokenList.front(); tok; tok = tok->next()) {
+                tok->clearAst();
+                tok->clearValueFlow();
+            }
+        }
+
+        mTokenList.createAst();
+        mTokenList.validateAst();
+
         // Locate possible instantiations of templates..
         getTemplateInstantiations();
 
diff --git a/lib/tokenlist.cpp b/lib/tokenlist.cpp
index df754288c..8ee5c7611 100644
--- a/lib/tokenlist.cpp
+++ b/lib/tokenlist.cpp
@@ -1655,9 +1655,17 @@ void TokenList::validateAst() const
         }
 
         // Don't check templates
-        if (tok->str() == "<" && tok->link()) {
-            tok = tok->link();
-            continue;
+        if (tok->str() == "<") {
+            if (tok->link()) {
+                tok = tok->link();
+                continue;
+            } else {
+                const Token *end = tok->findClosingBracket();
+                if (end) {
+                    tok = end;
+                    continue;
+                }
+            }
         }
 
```